Mage.Sets/src/mage/cards/a/AnrakyrTheTraveller.java

```diff
@@ -0,0 +1,198 @@
+        return new AnrakyrTheTraveller(this);
+    }
+}
+
+class AnrakyrTheTravellerEffect extends OneShotEffect {
+    private final FilterCard filter;
+
+    public AnrakyrTheTravellerEffect(FilterCard filter) {
+        super(Outcome.AIDontUseIt);
+        this.filter = filter;
+        this.staticText = "you may cast an " + filter.getMessage() + " from your hand or graveyard by paying life equal to its mana value rather than paying its mana cost.";
+    }
+
+    protected AnrakyrTheTravellerEffect(final AnrakyrTheTravellerEffect effect) {
+        super(effect);
+        this.filter = effect.filter;
+    }
+
+    
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+        
+        Set<Card> cards = player.getHand().getCards(filter, game);
+        cards.addAll(player.getGraveyard().getCards(filter, game));
+        
+        Map<UUID, List<Card>> cardMap = new HashMap<>();
+        for (Card card : cards) {
+            List<Card> castableComponents = getCastableComponents(card, filter, source, player, game, false);
+            if (!castableComponents.isEmpty()) {
+                cardMap.put(card.getId(), castableComponents);
+            }
+        }      
+        Card cardToCast;
+        switch (cardMap.size()) {
+            case 0:
+                return false;
+            default:
+                Cards castableCards = new CardsImpl(cardMap.keySet());
+                TargetCard target = new TargetCard(0, 1, Zone.ALL, filter);
+                target.withNotTarget(true);
+                player.choose(Outcome.Benefit, castableCards, target, source, game);
+                cardToCast = castableCards.get(target.getFirstTarget(), game);
+        }  
+
+        if (cardToCast == null) {
+            return false;
+        }
+
+
+        List<Card> partsToCast = cardMap.get(cardToCast.getId());
+        String partsInfo = partsToCast
+                .stream()
+                .map(MageObject::getIdName)
+                .collect(Collectors.joining(" or "));
+        if (cardToCast == null
+                || partsToCast.size() < 1
+                || !player.chooseUse(
+                Outcome.PlayForFree, "Cast spell by paying life equal to its mana value rather than paying its mana cost (" + partsInfo + ")?", source, game
+        )) {
+            return false;
+        }
+        partsToCast.forEach(card -> game.getState().setValue("PlayFromNotOwnHandZone" + card.getId(), Boolean.TRUE));
+        
+        // pay life
+        PayLifeCost lifeCost = new PayLifeCost(cardToCast.getSpellAbility().getManaCosts().manaValue());
+        Costs<Cost> newCosts = new CostsImpl<>();
+        newCosts.add(lifeCost);
+        newCosts.addAll(cardToCast.getSpellAbility().getCosts());
+        player.setCastSourceIdWithAlternateMana(cardToCast.getId(), null, newCosts);
```

Comment:
This is working well from my understanding (and I did test it live to confirm).

Took me a little time to wrap my head around, so detailing why if someone else is interested:
The player's alternative cast is reset often, and rebuild from `AsThough` Continuous Effects from there.
Very local change like that is working as the cast is right after and there is no reset of the  player's computed data/reapply of Continuous Effects from base objects.